### PR TITLE
bin: clean up flags and configurations parsing

### LINKED ISSUE #1299: logs/stderr: run tikv without arguments, it shows a stack backtrace.
Please answer these questions before submitting your issue. Thanks!

1. What version of Rust are you using (`rustc --version`)?

rustc 1.14.0-nightly (5665bdf3e 2016-11-02)

2. What operating system and processor architecture are you using?

macOS 10.11.6

3. What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.

./bin/tikv-server

4. What did you expect to see?

run with default configuration or exit with error code.

5. What did you see instead?
exit with error code 1 and got a stack backtrace.
```
2016-11-10 21:09:54,312 mod.rs:400 - INFO  - Welcome to TiKV.
2016-11-10 21:09:54,313 mod.rs:401 - INFO  - Version:
2016-11-10 21:09:54,313 mod.rs:402 - INFO  - Git Commit Hash: 936abf05de577f93c13d2725cec869f6c98a8e07
2016-11-10 21:09:54,313 mod.rs:403 - INFO  - UTC Build Time:  2016-11-10 12:59:26
2016-11-10 21:09:54,313 tikv-server.rs:83 - INFO  - server.addr, use default Some("127.0.0.1:20160")
2016-11-10 21:09:54,313 tikv-server.rs:828 - INFO  - Start listening on 127.0.0.1:20160...
2016-11-10 21:09:54,313 tikv-server.rs:83 - INFO  - pd.endpoints, use default None
2016-11-10 21:09:54,316 panic_hook.rs:85 - ERROR - thread 'main' panicked 'please specify pd.endpoints' at "../src/libcore/option.rs:705"
stack backtrace:
   0:        0x10608278e - backtrace::backtrace::trace::h16372ee7bf1517e5
   1:        0x106082a7c - backtrace::capture::Backtrace::new::hcc43c50c4b11c693
   2:        0x10606f7d4 - tikv::util::panic_hook::set_exit_hook::{{closure}}::hb14d8cbcd2b90ee2
   3:        0x1063564c3 - std::panicking::rust_panic_with_hook::h00b81bb3dcbd51f2
   4:        0x106356394 - std::panicking::begin_panic::ha6a0d553db9869ff
   5:        0x1063562b2 - std::panicking::begin_panic_fmt::h24d113aee3ee4081
   6:        0x106356217 - rust_begin_unwind
   7:        0x106382de0 - core::panicking::panic_fmt::he441b2ea2036b98a
   8:        0x106382e4d - core::option::expect_failed::h7136772e5b75d3f8
   9:        0x105f57587 - tikv_server::main::h000b13565d60ecd3
  10:        0x106357d8a - __rust_maybe_catch_panic
  11:        0x106356896 - std::rt::lang_start::h74b3afbdd8daef1c
```
Question:

Is a stack backtrace suitable here? Or just exit with last error log is enough.

### LINKED ISSUE #1300: bin: cleanup flag and config file parsing
Now we get configuration from command flags and toml file. 

E,g, We use `get_string_value` to get the string configuration, the function definition is:

```
fn get_string_value<F>(short: &str,
                       long: &str,
                       matches: &Matches,
                       config: &toml::Value,
                       default: Option<String>,
                       f: F)
``` 

As we see, we must pass `short`/`Matches`  for command flags and `long`/`config` for config file.
the `f` is for config file, and if we don't find anything, we will use the `default`.

This function is too complex. 

1. We have many configurations defined in config file, only have little in command flags. So we can handle flags and config file separately.
2. For config file, we don't need the `f`. If we want to get string, we can convert got value to string directly, no need `|v| v.as_str().map(|s| s.to_owned()` anymore. 
3. The configuration in command flags will overwrite the configuration in toml file if we set both, so the default value seems always for config file. If we pass a command flag, we will use this flag value, if not, we will try to use the config file value, else we will try to use the default value. 

So the functions look:

```
fn get_flag_int(matches: &Matches, name: &str) -> Option<int64>;
fn get_toml_int(config: &toml::Value, name: &str, default: Option<int64>) -> Option<int64>;
```

Btw, we must also fix #1299 too. I think for command flag parsing, if we meet error, we can even output help message too.

### PR DESCRIPTION
**Pull request by overvenus**
This PR replaces `get_(string|integer|boolean)_value` with `get_(toml|flag)_(string|int|boolean)`.

Now TiKV exits with `1` intead of panic when some important values are missing.

Close #1300 
Close #1299 

PTAL @siddontang @zhangjinpeng1987 @BusyJay

### CODE REVIEW

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -62,139 +63,107 @@ fn print_usage(program: &str, opts: Options) {
...
-                .or_else(|| {
-                    info!("{}, use default {:?}", long, default);
-                    default
-                })
+    s
+}
+
+fn get_toml_int(config: &toml::Value, name: &str, default: Option<i64>) -> i64 {
+    let i = config.lookup(name)
+        .and_then(|v| v.as_integer())
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I think readable int should be supported here.

File: src/bin/tikv-server.rs [hunk 2]
```diff
@@ -62,139 +63,107 @@ fn print_usage(program: &str, opts: Options) {
...
-                default
-            })
+        return i;
+    }
+    None
+}
+
+fn get_toml_boolean(config: &toml::Value, name: &str, default: Option<bool>) -> bool {
+    let b = config.lookup(name)
+        .and_then(|v| v.as_bool())
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: What if I accidentally set a string when bool is expected?

File: src/bin/tikv-server.rs [hunk 3]
```diff
@@ -62,139 +63,107 @@ fn print_usage(program: &str, opts: Options) {
...
-                        -> bool
-    where F: Fn(&toml::Value) -> Option<bool>
-{
-    let b = if matches.opt_defined(short) {
-        matches.opt_str(short).map(|x| x.parse::<bool>().unwrap())
-    } else {
-        None
-    };
+fn get_flag_int(matches: &Matches, name: &str) -> Option<i64> {
+    if matches.opt_defined(name) {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: I think this should be always true.

File: src/bin/tikv-server.rs (hunk 2 above)
Comment (overvenus) [inline src/bin/tikv-server.rs]: Then `v.as_bool() ` returns `None`, it tries to unwrap `default`. If `default` is `None` too, it exits with 1.

Comment (BusyJay) [inline src/bin/tikv-server.rs]: Why not just panic?

File: src/bin/tikv-server.rs [hunk 4]
```diff
@@ -62,139 +63,106 @@ fn print_usage(program: &str, opts: Options) {
...
-{
-    // avoid panic if short is not defined.
-    let s = if matches.opt_defined(short) {
-        matches.opt_str(short)
-    } else {
-        None
-    };
+fn get_flag_string(matches: &Matches, name: &str) -> Option<String> {
+    let s = matches.opt_str(name);
+    info!("flag {}: {:?}", name, s);
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Use `debug` instead.

File: src/bin/tikv-server.rs [hunk 5]
```diff
@@ -62,139 +63,106 @@ fn print_usage(program: &str, opts: Options) {
...
+fn get_flag_int(matches: &Matches, name: &str) -> Option<i64> {
+    let i = matches.opt_str(name).map(|x| {
+        x.parse::<i64>()
+            .or_else(|_| util::config::parse_readable_int(&x))
+            .unwrap_or_else(|e| {
+                warn!("parse {} failed: {:?}", name, e);
+                process::exit(1)
             })
+    });
+    info!("flag {}: {:?}", name, i);
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Ditto.

File: src/bin/tikv-server.rs [hunk 6]
```diff
@@ -62,139 +63,106 @@ fn print_usage(program: &str, opts: Options) {
...
+        .or_else(|| {
+            info!("{}, use default {:?}", name, default);
+            default
         })
-        .expect(&format!("please specify {}", long))
+        .unwrap_or_else(|| {
+            warn!("please specify {}", name);
+            process::exit(1)
+        });
+    info!("toml value {}: {:?}", name, b);
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Ditto.

File: src/bin/tikv-server.rs (hunk 2 above)
Comment (siddontang) [inline src/bin/tikv-server.rs]: To avoid panic.

I think we should output error here and exit directly.

File: src/bin/tikv-server.rs (hunk 4 above)
Comment (siddontang) [inline src/bin/tikv-server.rs]: we should output the flag here, so we can know what user passes.

File: src/bin/tikv-server.rs [hunk 7]
```diff
@@ -65,7 +65,7 @@ fn print_usage(program: &str, opts: Options) {
 
 fn get_flag_string(matches: &Matches, name: &str) -> Option<String> {
     let s = matches.opt_str(name);
-    info!("flag {}: {:?}", name, s);
+    debug!("flag {}: {:?}", name, s);
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: I prefer outputting the command flags here with INFO level, so we can know what really user passes.  

/cc @ngaut

Comment (ngaut) [inline src/bin/tikv-server.rs]: I do agree with you @siddontang

File: src/bin/tikv-server.rs [hunk 8]
```diff
@@ -62,139 +63,108 @@ fn print_usage(program: &str, opts: Options) {
...
-                       -> String
-    where F: Fn(&toml::Value) -> Option<String>
-{
-    // avoid panic if short is not defined.
-    let s = if matches.opt_defined(short) {
-        matches.opt_str(short)
-    } else {
-        None
-    };
+fn exit(code: i32, msg: String) -> ! {
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: I see that the code is already 1, why not changing to exit_with_err(msg: String) instead?

**Comment by ngaut** [comment]
Ping @overvenus

**Comment by overvenus** [comment]
@zhangjinpeng1987 PTAL

This PR changes a lot of rocksdb configurations.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/bin/tikv-server.rs [hunk 9]
```diff
@@ -62,139 +63,108 @@ fn print_usage(program: &str, opts: Options) {
...
+    s
+}
+
+fn get_toml_int(config: &toml::Value, name: &str, default: Option<i64>) -> i64 {
+    let i = match config.lookup(name) {
+            Some(&toml::Value::Integer(i)) => Some(i),
+            Some(&toml::Value::String(ref s)) => {
+                util::config::parse_readable_int(s)
+                    .map_err(|e| {
+                        info!("{} parse failed {:?}", name, e);
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Why not just `exit_with_err`?

File: src/bin/tikv-server.rs [hunk 10]
```diff
@@ -62,139 +63,108 @@ fn print_usage(program: &str, opts: Options) {
...
-    where F: Fn(&toml::Value) -> Option<String>
-{
-    // avoid panic if short is not defined.
-    let s = if matches.opt_defined(short) {
-        matches.opt_str(short)
-    } else {
-        None
-    };
+fn exit_with_err(msg: String) -> ! {
+    warn!("{}", msg);
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: Use `error!` instead.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**